pkg/queue/cluster_queue_impl.go

```diff
@@ -55,39 +72,99 @@ func (c *ClusterQueueImpl) Cohort() string {
 func (c *ClusterQueueImpl) AddFromQueue(q *Queue) bool {
 	added := false
 	for _, info := range q.items {
-		if c.pushIfNotPresent(info) {
+		if c.heap.PushIfNotPresent(info) {
 			added = true
 		}
 	}
 	return added
 }
 
+func (c *ClusterQueueImpl) PushOrUpdate(wInfo *workload.Info) {
+	key := workload.Key(wInfo.Obj)
+	oldInfo := c.inadmissibleWorkloads[key]
+	if oldInfo != nil {
+		// update in place if the workload was inadmissible and didn't change
+		// to potentially become admissible.
+		if equality.Semantic.DeepEqual(oldInfo.Obj.Spec, wInfo.Obj.Spec) {
+			c.inadmissibleWorkloads[key] = wInfo
+			return
+		}
+		// otherwise move or update in place in the queue.
+		delete(c.inadmissibleWorkloads, key)
+	}
+	c.heap.PushOrUpdate(wInfo)
+}
+
+func (c *ClusterQueueImpl) Delete(w *kueue.Workload) {
+	delete(c.inadmissibleWorkloads, workload.Key(w))
+	c.heap.Delete(workload.Key(w))
+}
+
 func (c *ClusterQueueImpl) DeleteFromQueue(q *Queue) {
+	for _, w := range q.items {
+		key := workload.Key(w.Obj)
+		if wl := c.inadmissibleWorkloads[key]; wl != nil {
+			delete(c.inadmissibleWorkloads, key)
+		}
+	}
 	for _, w := range q.items {
 		c.Delete(w.Obj)
 	}
 }
 
-// pushIfNotPresent pushes the workload to ClusterQueue.
-// If the workload is already present, returns false. Otherwise returns true.
-func (c *ClusterQueueImpl) pushIfNotPresent(info *workload.Info) bool {
-	return c.heap.PushIfNotPresent(info)
-}
+// RequeueIfNotPresent inserts a workload that cannot be admitted into
+// ClusterQueue, unless it is already in the queue. If immediate is true,
+// the workload will be pushed back to heap directly. If not,
+// the workload will be put into the inadmissibleWorkloads.
+func (c *ClusterQueueImpl) RequeueIfNotPresent(wInfo *workload.Info, immediate bool) bool {
```

Comment:
In StrictFIFO, we should only check 1)If the namespaceSelector match, push into heap 2)If not match, push into inadmissibleWorkloads. So parameter `immediate` is not necessary. So, we should implement this logic in `ClusterQueueImpl`, StrictFIFO should just call the function, For BestEffortFIFO, it should depend on the `immediate` then.  Also, we can remove the `RequeueReason`.

Correct me if I misunderstand the codes.

Reply:
the check on namespaceSelector happens outside the RequeueIfNotPresent function during the scheduling loop, so we need to pass that down, and the way we did it is via the `RequeueReason`; immediate is now an implementation detail of the interface